### ISSUE TITLE
Support for array-wise operation in eval_rows function

### ISSUE THREAD

**Issue by HxHgk** [jpmml/sklearn2pmml#308]
Currently, the `eval_rows` function (shown below) in "sklearn2pmml/util/\_\_init\_\_.py" applies element-wise for an input `X` which does not have an `apply` attribute, for example when `X` is a ndarray. This can be really slow and reduces the speed of relevant operations like `LookupTransformer`. 
Is it possible to support array-wise vectorized computation here?
```python
def eval_rows(X, func, dtype = object):
	if hasattr(X, "apply"):
		return X.apply(func, axis = 1)
	nrow = X.shape[0]
	Xt = numpy.empty(shape = (nrow, ), dtype = dtype)
	for i in range(0, nrow):
		Xt[i] = func(X[i])
	return Xt
```

**Comment by vruusmann** [comment]
> Is it possible to support array-wise vectorized computation here?

Show me how, and it can be implemented.

The `eval_rows` util function is a general purpose one. If you need to speed up specific transformers (eg. `LookupTransformer`), then the effect should come from avoiding an `eval_rows` function invocation altogether. In other words, the vectorization for dict lookups should happen already inside `LookupTransformer` and use some very specific vectorized operation.

**Comment by vruusmann** [comment]
Reopening, because the lifecycle of GitHub issues isn't quite like that.

TLDR: If you have ideas how to make individual transformations (eg. `LookupTransformer` better), then please share the code.

**Comment by HxHgk** [comment]
Sorry if I caused any confusion!
Here's the reason I raised this issue. My code ran really really slow and I pinned down the problem to the transform function in `LookupTransformer` (quoted below). I didn't think anything could go wrong with the dictionary and thought the bottleneck was the element-wise `eval_rows`. 
Later, I realized the real issue was actually with the dictionary because I messed up the dtype of the key. `eval_rows` was innocent in this case, and I thus closed the issue. I should've left a comment to explain. Sorry for the confusion again!
Honestly, I think element-wise operation is unavoidable in some sense to keep the application of `eval_rows` general. I don't have a good idea of how to vectorize it yet.
```python
	def transform(self, X):
		X = column_or_1d(X, warn = True)
		transform_dict = self._transform_dict()
		func = lambda k: transform_dict[k]
		Xt = eval_rows(X, func)
		return _col2d(Xt)
```

**Comment by vruusmann** [comment]
> Later, I realized the real issue was actually with the dictionary because I messed up the dtype of the key.

Interesting. Perhaps the `LookupTransformer` class should perform an additional sanity check, to ensure that the datatype of dict keys and `X` are the same?

Did you know that you can use `sklearn2pmml.preprocessing.CastTransformer` to "refine" the data type (in the middle of a pipeline)?

**Comment by HxHgk** [comment]
For my case, it's not due to the mismatch of datatype between the input `X` and the dict key. Instead, I messed up the key while constructing the dictionary outside of `LookupTransformer`. So a sanity check inside `LookupTransformer` would not catch my issue. It might be good for other cases though. 
Good to know `CastTransformer`. Thanks!
